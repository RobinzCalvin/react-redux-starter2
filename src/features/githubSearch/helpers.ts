export function injectSizeToAvatarURL(URL: string, sizeInPixels: number) {
  return `${URL}&s=${sizeInPixels}`;
}
