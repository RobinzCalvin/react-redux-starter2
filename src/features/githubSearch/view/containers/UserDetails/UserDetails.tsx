import * as React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { IAppReduxState } from 'shared/types/app';
import { IUser } from 'shared/types/models';
import { Dialog } from 'shared/view/components';

import { actions, selectors } from './../../../redux';
import './UserDetails.scss';

interface IOwnProps {
  userLogin: string | null; // think of login or username everywhere
  onClose(): void;
}

interface IStateProps {
  userDetails: IUser | null;
}

interface IActionProps {
  loadUserDetails: typeof actions.loadUserDetails;
}

type IProps = IStateProps & IActionProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    userDetails: selectors.selectUserDetails(state),
  };
}

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    loadUserDetails: actions.loadUserDetails,
  }, dispatch);
}

const b = block('user-details');

class UserDetails extends React.PureComponent<IProps> {
  public render() {
    const { userLogin } = this.props;
    return (
      <Dialog
        open={userLogin !== null}
        title="User details"
        onEnter={this.handleDialogEnter}
        onClose={this.handleDialogClose}
      >
        <div className={b()}>content</div>
      </Dialog>
    );
  }

  @bind
  private handleDialogEnter() {
    const { userLogin, loadUserDetails } = this.props;
    if (userLogin) {
      loadUserDetails(userLogin);
    }
  }

  @bind
  private handleDialogClose() {
    const { onClose } = this.props;
    onClose();
  }
}

export default connect(mapState, mapDispatch)(UserDetails);
