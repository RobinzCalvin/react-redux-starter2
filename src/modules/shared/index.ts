export { default as Layout } from './Layout/Layout';
export { default as withLayout } from './withLayout';
