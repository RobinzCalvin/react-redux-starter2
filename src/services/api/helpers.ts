import { IUserSearchOptions } from 'shared/types/github';

function constructReposNumberQuery(minRepos?: number, maxRepos?: number) {
  if (maxRepos === void 0 && (minRepos && minRepos > 0)) {
    return `+repos:>${minRepos}`;
  } else if (minRepos === void 0 && (maxRepos && maxRepos > 0)) {
    return `+repos:<${maxRepos}`;
  } else if (minRepos && maxRepos) {
    return `+repos:${minRepos}..${maxRepos}`;
  }
  return '';
}

export function constructUserSearchQuery(queryString: string, options: IUserSearchOptions, page: number = 1) {
  const { searchBy, searchType, minRepos, maxRepos, reposLanguage, perPage } = options;
  return queryString.concat(
    searchBy !== 'username-email' ? `+in:${searchBy}` : '',
    searchType !== 'both' ? `+type:${searchType}` : '',
    reposLanguage ? `+language:${reposLanguage}` : '',
    constructReposNumberQuery(minRepos, maxRepos),
    `&per_page=${perPage}`,
    `&page=${page}`,
  );
}

/*
  header format (requesting page=7):
  <https://api.github.com/search/users?q=asd&per_page=30&page=6>; rel="prev",
  <https://api.github.com/search/users?q=asd&per_page=30&page=8>; rel="next",
  <https://api.github.com/search/users?q=asd&per_page=30&page=34>; rel="last",
  <https://api.github.com/search/users?q=asd&per_page=30&page=1>; rel="first"
*/
export function getTotalPagesFromLinkHeader(link?: string): number {
  if (link !== void 0) {
    const lastPageMatch = link.match(/&page=(\d+)>; rel="last"/);
    if (lastPageMatch) {
      return Number(lastPageMatch[1]);
    } else {
      const prevPageMatch = link.match(/&page=(\d+)>; rel="prev"/);
      if (prevPageMatch) {
        return Number(prevPageMatch[1]) + 1;
      } else {
        console.error(`Error while trying to get total pages from ${link}`);
      }
    }
  }
  return 1;
}
