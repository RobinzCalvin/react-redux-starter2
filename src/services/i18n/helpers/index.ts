export * from './guards';
export * from './buildTranslationKeys';
