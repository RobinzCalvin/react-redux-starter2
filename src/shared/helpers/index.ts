export { default as withComponent } from './withComponent';
