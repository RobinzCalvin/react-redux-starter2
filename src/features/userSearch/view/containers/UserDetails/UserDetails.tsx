import * as React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { IAppReduxState } from 'shared/types/app';
import { IDetailedUser } from 'shared/types/models';
import { Dialog } from 'shared/view/components';
import { Typography, Preloader } from 'shared/view/elements';

import { injectSizeToAvatarURL } from '../../../helpers';
import { actions, selectors } from './../../../redux';
import './UserDetails.scss';

interface IOwnProps {
  username: string | null;
  onClose(): void;
}

interface IStateProps {
  userDetails: IDetailedUser | null;
  isLoadUserDetailsRequesting: boolean;
}

interface IActionProps {
  loadUserDetails: typeof actions.loadUserDetails;
}

type IProps = IStateProps & IActionProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    userDetails: selectors.selectUserDetails(state),
    isLoadUserDetailsRequesting: selectors.selectCommunication(state, 'loadUserDetails').isRequesting,
  };
}

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    loadUserDetails: actions.loadUserDetails,
  }, dispatch);
}

const b = block('user-details');

class UserDetails extends React.PureComponent<IProps> {
  private avatarSize = 230;

  public render() {
    const { username, isLoadUserDetailsRequesting } = this.props;
    return (
      <Dialog
        open={username !== null}
        title="User details"
        onEnter={this.handleDialogEnter}
        onClose={this.handleDialogClose}
      >
        <div className={b()}>
          <Preloader size={80} isShown={isLoadUserDetailsRequesting} backgroundColor="#fff" />
          {this.renderContent()}
        </div>
      </Dialog>
    );
  }

  private renderContent() {
    const { userDetails } = this.props;
    if (userDetails) {
      return (
        <>
          {this.renderMain(userDetails)}
          {this.renderAttributes(userDetails)}
        </>
      );
    }
    return null;
  }

  private renderMain(userDetails: IDetailedUser) {
    const { htmlURL, avatarURL, realName, username, location } = userDetails;
    return (
      <a href={htmlURL} className={b('main')} target="_blank">
        <img
          className={b('avatar')}
          src={injectSizeToAvatarURL(avatarURL, this.avatarSize)}
        />
        <Typography variant="h5">{realName}</Typography>
        <Typography variant="subtitle1">{username}</Typography>
        {location && <Typography variant="subtitle2" color="textSecondary">{location}</Typography>}
      </a>
    );
  }

  private renderAttributes(userDetails: IDetailedUser) {
    const { followersNumber, followingNumber, reposNumber, htmlURL } = userDetails;
    return (
      <div className={b('attributes')}>
        <a href={`${htmlURL}/followers`} target="_blank" className={b('attribute')}>
          Followers
          <span className={b('value')}>{followersNumber}</span>
        </a>
        <a href={`${htmlURL}/following`} target="_blank" className={b('attribute')}>
          Following
          <span className={b('value')}>{followingNumber}</span>
        </a>
        <a href={`${htmlURL}/repositories`} target="_blank" className={b('attribute')}>
          Repositories
          {/* TODO: тут можно контейнер провайдером загрузку реп добавить (репы будут справа выезжать по стрелочке) */}
          <span className={b('value')}>{reposNumber}</span>
        </a>
        </div>
    );
  }

  @bind
  private handleDialogEnter() {
    const { username, loadUserDetails } = this.props;
    if (username) {
      loadUserDetails(username);
    }
  }

  @bind
  private handleDialogClose() {
    const { onClose } = this.props;
    onClose();
  }
}

export default connect(mapState, mapDispatch)(UserDetails);
