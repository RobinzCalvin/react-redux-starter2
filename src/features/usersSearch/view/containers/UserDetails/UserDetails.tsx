import * as React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { IAppReduxState } from 'shared/types/app';
import { IDetailedUser } from 'shared/types/models';
import { Dialog } from 'shared/view/components';
import { Typography, Preloader } from 'shared/view/elements';

import { injectSizeToAvatarURL } from '../../../helpers';
import { actions, selectors } from './../../../redux';
import './UserDetails.scss';

interface IOwnProps {
  username: string;
  onClose(): void;
}

interface IStateProps {
  userDetails: IDetailedUser | null;
  isLoadUserDetailsRequesting: boolean;
}

interface IActionProps {
  loadUserDetails: typeof actions.loadUserDetails;
}

type IProps = IStateProps & IActionProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    userDetails: selectors.selectUserDetails(state),
    isLoadUserDetailsRequesting: selectors.selectCommunication(state, 'loadUserDetails').isRequesting,
  };
}

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    loadUserDetails: actions.loadUserDetails,
  }, dispatch);
}

const b = block('user-details');

class UserDetails extends React.PureComponent<IProps> {
  private avatarSize = 230;

  public render() {
    const { isLoadUserDetailsRequesting } = this.props;
    return (
      <Dialog
        open={true}
        title="User details"
        onEnter={this.handleDialogEnter}
        onClose={this.handleDialogClose}
      >
        <div className={b()}>
          <Preloader size={80} isShown={isLoadUserDetailsRequesting} backgroundColor="#fff" />
          {this.renderContent()}
        </div>
      </Dialog>
    );
  }

  private renderContent() {
    const { userDetails } = this.props;
    if (userDetails) {
      return (
        <>
          {this.renderMain(userDetails)}
          {this.renderAttributes(userDetails)}
        </>
      );
    }
    return null;
  }

  private renderMain(userDetails: IDetailedUser) {
    const { htmlURL, avatarURL, realName, username, location } = userDetails;
    return (
      <a href={htmlURL} className={b('main')} target="_blank">
        <img
          className={b('avatar')}
          src={injectSizeToAvatarURL(avatarURL, this.avatarSize)}
        />
        <Typography variant="h5">{realName}</Typography>
        <Typography variant="subtitle1">{username}</Typography>
        {location && <Typography variant="subtitle2" color="textSecondary">{location}</Typography>}
      </a>
    );
  }

  private renderAttributes(userDetails: IDetailedUser) {
    const { followersNumber, followingNumber, reposNumber, htmlURL } = userDetails;
    return (
      <div className={b('attributes')}>
        {this.renderAttribute(`${htmlURL}/followers`, 'Followers', followersNumber)}
        {this.renderAttribute(`${htmlURL}/following`, 'Following', followingNumber)}
        {this.renderAttribute(`${htmlURL}/repositories`, 'Repositories', reposNumber)}
      </div>
    );
  }

  private renderAttribute(URL: string, title: string, value: number) {
    return (
      <a href={URL} target="_blank" className={b('attribute')}>
        {title}
        <span className={b('value')}>{value}</span>
      </a>
    );
  }

  @bind
  private handleDialogEnter() {
    const { username, loadUserDetails } = this.props;
    if (username) {
      loadUserDetails(username);
    }
  }

  @bind
  private handleDialogClose() {
    const { onClose } = this.props;
    onClose();
  }
}

export default connect(mapState, mapDispatch)(UserDetails);
