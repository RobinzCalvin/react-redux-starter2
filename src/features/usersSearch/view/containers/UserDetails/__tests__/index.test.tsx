import React from 'react';
import { shallow } from 'enzyme';

import { getDetailedGithubUserMock } from 'shared/helpers';

import { UserDetails, IUserDetailsProps } from '../UserDetails';

const props: IUserDetailsProps = {
  userDetails: getDetailedGithubUserMock(),
  isLoadUserDetailsRequesting: false,
  loadUserDetails: jest.fn(),
  onClose: jest.fn(),
  username: 'the user',
};

describe('UserDetails component', () => {
  const component = shallow(<UserDetails {...props} />);
  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
