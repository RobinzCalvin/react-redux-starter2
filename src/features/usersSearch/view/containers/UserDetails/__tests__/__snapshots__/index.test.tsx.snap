// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails component should match snapshot 1`] = `
<Dialog
  onClose={[Function]}
  onEnter={[Function]}
  open={true}
  title="User details"
>
  <Component>
    <div
      className="user-details"
    >
      <Preloader
        backgroundColor="#fff"
        isShown={false}
        size={80}
      />
      <a
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        href="http://the-url.com"
        target="_blank"
      >
        <img
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          src="http://the-url.com&s=230"
        />
        <WithStyles(Typography)
          variant="h5"
        >
          The real name
        </WithStyles(Typography)>
        <WithStyles(Typography)
          variant="subtitle1"
        >
          username
        </WithStyles(Typography)>
        <WithStyles(Typography)
          color="textSecondary"
          variant="subtitle2"
        >
          Nowhere
        </WithStyles(Typography)>
      </a>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <a
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          href="http://the-url.com/followers"
          target="_blank"
        >
          Followers
          <span
            className={
              Object {
                "has": [Function],
                "is": [Function],
                "mix": [Function],
                "split": [Function],
                "state": [Function],
                "toString": [Function],
              }
            }
          >
            12
          </span>
        </a>
        <a
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          href="http://the-url.com/following"
          target="_blank"
        >
          Following
          <span
            className={
              Object {
                "has": [Function],
                "is": [Function],
                "mix": [Function],
                "split": [Function],
                "state": [Function],
                "toString": [Function],
              }
            }
          >
            12
          </span>
        </a>
        <a
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          href="http://the-url.com/repositories"
          target="_blank"
        >
          Repositories
          <span
            className={
              Object {
                "has": [Function],
                "is": [Function],
                "mix": [Function],
                "split": [Function],
                "state": [Function],
                "toString": [Function],
              }
            }
          >
            12
          </span>
        </a>
      </div>
    </div>
  </Component>
</Dialog>
`;
